src/utils/internalLinks.js

```diff
@@ -181,39 +200,58 @@ export const getNarrowFromLink = (
       //   group PM conversation excludes self, so it's unusable for anyone
       //   else.  In particular this will foil you if, say, you try to give
       //   someone else in the conversation a link to a particular message.
-      const ids = parsePmOperand(paths[1]);
+      const ids = parsePmOperand(hashSegments[1]);
       return pmNarrowFromRecipients(pmKeyRecipientsFromIds(ids, ownUserId));
     }
     case 'topic': {
-      const stream = parseStreamOperand(paths[1], streamsById, streamsByName);
-      return stream && topicNarrow(stream.stream_id, parseTopicOperand(paths[3]));
+      const stream = parseStreamOperand(hashSegments[1], streamsById, streamsByName);
+      return stream && topicNarrow(stream.stream_id, parseTopicOperand(hashSegments[3]));
     }
     case 'stream': {
-      const stream = parseStreamOperand(paths[1], streamsById, streamsByName);
+      const stream = parseStreamOperand(hashSegments[1], streamsById, streamsByName);
       return stream && streamNarrow(stream.stream_id);
     }
     case 'special':
       try {
-        return specialNarrow(paths[1]);
+        return specialNarrow(hashSegments[1]);
       } catch {
         return null;
       }
+    case 'home':
+      return null; // TODO(?): Give HOME_NARROW
     default:
+      ensureUnreachable(type);
       return null;
   }
 };
 
 /**
- * TODO write jsdoc
+ * From a URL and realm with `isNarrowLink(url, realm) === true`, give
+ *   message_id if the URL has /near/{message_id}, otherwise give zero.
  *
  * This performs a call to `new URL` and therefore may take a fraction of a
  * millisecond.  Avoid using in a context where it might be called more than
  * 10 or 100 times per user action.
  */
 export const getMessageIdFromLink = (url: string, realm: URL): number => {
-  const paths = getPathsFromUrl(url, realm);
+  // isNarrowLink(…) is true, by jsdoc, so this call is OK.
+  const hashSegments = getHashSegmentsFromNarrowLink(url, realm);
+
+  // This and nearOperandIndex can simplify when we rename/repurpose
+  //   getHashSegmentsFromNarrowLink so it gives an array of
+  //   { negated, operator, operand } objects; see TODO there.
+  const nearOperatorIndex = hashSegments.findIndex(
+    (str, i) =>
+      // This is a segment where we expect an operator to be specified.
+      i % 2 === 0
+      // The operator is 'near' and its meaning is not negated (`str` does
+      // not start with "-").
+      && str === 'near',
+  );
+  // We expect an operand to directly follow its operator.
+  const nearOperandIndex = nearOperatorIndex + 1; // ignore if "near" not found
 
-  return isMessageLink(url, realm) ? parseInt(paths[paths.lastIndexOf('near') + 1], 10) : 0;
+  return nearOperatorIndex >= 0 ? parseInt(hashSegments[nearOperandIndex], 10) : 0;
```

Comment:
nit: rather than a comment where the variable is defined, saying it's only meaningful in certain cases, it's more robust to arrange things so the variable doesn't get defined in the first place when the value won't be meaningful.

So e.g.
```js
  if (nearOperatorIndex < 0) {
    return 0;
  }

  // We expect an operand to directly follow its operator.
  const nearOperandIndex = nearOperatorIndex + 1;
  return parseInt(hashSegments[nearOperandIndex], 10);
```

(Doesn't *really* matter in this instance because this variable only remains in scope for a couple of lines -- there isn't much space in which to forget that caveat on it. But it's a better pattern in general. … Also, rearranging this way highlights how this function is returning 0 in the "doesn't exist" case. That'd be good to change to `null` before actually using this function's return value, though that doesn't need to be within this PR.)